From 05f148f30f4c95c988099a76977a56f546707215 Mon Sep 17 00:00:00 2001
From: Zhong Yichong <zyc@rock-chips.com>
Date: Fri, 2 Feb 2018 09:25:38 +0800
Subject: [PATCH] camera: rockchip: camsys_drv: v0.0x26.0

fix iommu resource not been released when process mediaserver
crashes unexpectly.

Change-Id: Ia8209f7d0a60f6a86d273e313260b87d5facecc3
Signed-off-by: Zhong Yichong <zyc@rock-chips.com>
---
 drivers/media/video/rk_camsys/camsys_drv.c     | 18 ++++++++++++++++++
 .../media/video/rk_camsys/camsys_internal.h    |  6 +++++-
 drivers/media/video/rk_camsys/camsys_marvin.c  |  9 ++++++---
 3 files changed, 29 insertions(+), 4 deletions(-)

diff --git a/drivers/media/video/rk_camsys/camsys_drv.c b/drivers/media/video/rk_camsys/camsys_drv.c
index 94a07edfd93b..ba47c656ed62 100644
--- a/drivers/media/video/rk_camsys/camsys_drv.c
+++ b/drivers/media/video/rk_camsys/camsys_drv.c
@@ -848,6 +848,9 @@ static int camsys_release(struct inode *inode, struct file *file)
 {
 	camsys_dev_t *camsys_dev = (camsys_dev_t *)file->private_data;
 	unsigned int i, phycnt;
+	camsys_sysctrl_t devctl;
+	camsys_iommu_t *iommu_par;
+	int index;
 
 	camsys_irq_disconnect(NULL, camsys_dev, true);
 
@@ -860,6 +863,21 @@ static int camsys_release(struct inode *inode, struct file *file)
 			}
 		}
 	}
+	/*
+	 * iommu resource might not been released when mediaserver process
+	 * died unexpectly, so we force to release the iommu resource here
+	 */
+	devctl.dev_mask = camsys_dev->dev_id;
+	devctl.ops = CamSys_IOMMU;
+	devctl.on = 0;
+	iommu_par = (camsys_iommu_t *)(devctl.rev);
+	/* used as force release flag */
+	iommu_par->client_fd = -1;
+	for (index = 0; index < CAMSYS_DMA_BUF_MAX_NUM ; index++) {
+		if (camsys_dev->iommu_cb(camsys_dev, &devctl) != 0)
+			break;
+	}
+
 	atomic_dec(&camsys_dev->refcount);
 	camsys_trace(1,
 		"%s(%p) is closed",
diff --git a/drivers/media/video/rk_camsys/camsys_internal.h b/drivers/media/video/rk_camsys/camsys_internal.h
index 78672437336b..63f06803697c 100644
--- a/drivers/media/video/rk_camsys/camsys_internal.h
+++ b/drivers/media/video/rk_camsys/camsys_internal.h
@@ -192,9 +192,13 @@
        1) function is the same as commit in v0.0x22.3 but now is better way.
 *v0.0x25.0:
 	1) support px30.
+*v0.0x26.0:
+       1) v0.0x21.9 may not fix all the case of iommu issue caused by the
+       unexpected termination of process cameraserver, so we force to release
+       all iommu resource in |.release| of fops aganin if needed.
 */
 
-#define CAMSYS_DRIVER_VERSION                   KERNEL_VERSION(0, 0x25, 0)
+#define CAMSYS_DRIVER_VERSION                   KERNEL_VERSION(0, 0x26, 0)
 
 #define CAMSYS_PLATFORM_DRV_NAME                "RockChip-CamSys"
 #define CAMSYS_PLATFORM_MARVIN_NAME             "Platform_MarvinDev"
diff --git a/drivers/media/video/rk_camsys/camsys_marvin.c b/drivers/media/video/rk_camsys/camsys_marvin.c
index ffabb4f51ddb..bef1dc1536a9 100644
--- a/drivers/media/video/rk_camsys/camsys_marvin.c
+++ b/drivers/media/video/rk_camsys/camsys_marvin.c
@@ -479,11 +479,14 @@ static int camsys_mrv_drm_iommu_cb(void *ptr, camsys_sysctrl_t *devctl)
 			(index >= CAMSYS_DMA_BUF_MAX_NUM))
 			return -EINVAL;
 
-		for (index = 0; index < camsys_dev->dma_buf_cnt; index++) {
-			if (camsys_dev->dma_buf[index].fd == iommu->map_fd)
+		for (index = 0; index < CAMSYS_DMA_BUF_MAX_NUM; index++) {
+			if (camsys_dev->dma_buf[index].fd == iommu->map_fd ||
+			    /* force release */
+			    (camsys_dev->dma_buf[index].fd != -1 &&
+			     iommu->client_fd == -1))
 				break;
 		}
-		if (index == camsys_dev->dma_buf_cnt) {
+		if (index == CAMSYS_DMA_BUF_MAX_NUM) {
 			camsys_warn("can't find map fd %d", iommu->map_fd);
 			return -EINVAL;
 		}
-- 
2.35.3

